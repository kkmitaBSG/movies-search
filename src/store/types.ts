export const api_key = "92b418e837b833be308bbfb1fb2aca1e";

export enum MoviesActionType {
  getMovies = "GET_MOVIES",
  getMoviesFail = "GET_MOVIES_FAIL",
  getSearchMovies = "GET_SEARCH_MOVIES",
  getMoviesNextPage = "GET_MOVIES_NEXT_PAGE",
  getMovieDetails = "GET_MOVIE_DETAILS",
}

export interface Movie {
  id: string;
  title: string;
  overview: string;
  poster_path?: string;
  poster?: string;
}

export interface Movies {
  results?: Movie[];
  page: number;
}

export interface MovieDetails {
  id?: string;
  backdrop_path?: string;
  title?: string;
  backdrop?: string;
  overview?: string;
  production_countries?: Array<{ name: string }>;
  release_date?: string;
  genres?: Array<{ name: string }>;
}

export interface MovieDetailsState {
  [id: string]: MovieDetails;
}

export interface State {
  movies: Movies;
  movieDetails: MovieDetails;
}

export const moviePosterOriginalPath = "https://image.tmdb.org/t/p/w500";
