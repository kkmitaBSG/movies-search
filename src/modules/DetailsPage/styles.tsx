import styled from "styled-components";

export const Poster = styled.img`
  width: 100%;
  max-height: 50vh;
  object-fit: cover;
  box-shadow: 0 0 200px black;
`;

export const InfoWrapper = styled.div`
  display: flex;
  margin: 60px;
  gap: 140px;
`;
export const LeftSide = styled.div`
  flex: 2;
`;
export const RightSide = styled.div`
  flex: 1;
`;

export const Title = styled.div`
  font-size: 45px;
  color: #ffbc9e;
`;
export const Description = styled.div`
  font-size: 20px;
  margin-top: 60px;
`;

export const AdditionalInfo = styled.div`
  font-size: 28px;
  display: flex;
  gap: 20px;
  margin-bottom: 40px;
`;
export const Name = styled.div`
  font-weight: 500;
  color: #ffbc9e;
`;

export const CustomLazyLoading = styled.div`
  wdth: 100%;
  height: 50vh;
  background: black;
`;