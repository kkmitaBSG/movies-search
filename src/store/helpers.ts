import { Movie, MovieDetails, moviePosterOriginalPath } from "./types";

export const transformMovieData = (movies?: Movie[]) =>
  movies?.map((movie) => ({
    ...movie,
    poster: moviePosterOriginalPath + movie.poster_path,
  }));

export const transformMovieDetailsData = (movieDetail?: MovieDetails) => ({
  ...movieDetail,
  backdrop: moviePosterOriginalPath + movieDetail?.backdrop_path,
});
