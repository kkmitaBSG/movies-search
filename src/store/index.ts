import { configureStore } from "@reduxjs/toolkit";
import popularMoviesReducer from "./reducers/popularMoviesReducer";
import movieDetailsReducer from "./reducers/movieDetailsReducer";
import createSagaMiddleware from "redux-saga";
import rootSaga from "./sagas";
import searchMoviesReducer from "./reducers/searchMoviesReducer";
import {
  createRouterMiddleware,
  createRouterReducerMapObject,
} from "@lagunovsky/redux-react-router";
import { createBrowserHistory } from "history";

export const history = createBrowserHistory();
const routerMiddleware = createRouterMiddleware(history);

const sagaMiddleware = createSagaMiddleware();

export const store = configureStore({
  reducer: {
    movies: popularMoviesReducer,
    searchMovies: searchMoviesReducer,
    movieDetails: movieDetailsReducer,
    ...createRouterReducerMapObject(history),
  },
  middleware: [sagaMiddleware, routerMiddleware],
});

sagaMiddleware.run(rootSaga);

export type RootState = ReturnType<typeof store.getState>;
export type AppDispatch = typeof store.dispatch;
