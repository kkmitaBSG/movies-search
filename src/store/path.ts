const API_PATH = `https://api.themoviedb.org/3`;

const MOVIE_PATH = `${API_PATH}/movie`;

export const MOVIE_POPULAR_PATH = `${MOVIE_PATH}/popular`;

export const MOVIE_DETAILS_PATH = `${MOVIE_PATH}/[id]`;

export const SEARCH_MOVIE_PATH = `${API_PATH}/search/movie`;
