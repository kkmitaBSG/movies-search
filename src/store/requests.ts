import axios from "axios";
import { api_key, Movie, Movies } from "./types";
import {
  MOVIE_DETAILS_PATH,
  MOVIE_POPULAR_PATH,
  SEARCH_MOVIE_PATH,
} from "./path";

export const getMovies = (pageNum: number) =>
  axios.get<Movies[]>(
    MOVIE_POPULAR_PATH + `?api_key=${api_key}&page=${pageNum}`
  );

export const getSearchMovies = (searchPhrase: string, pageNum: number) =>
  axios.get<Movies[]>(
    SEARCH_MOVIE_PATH +
      `?api_key=${api_key}&page=${pageNum}&query=${searchPhrase}`
  );

export const getMovieDetails = (movieId: string) =>
  axios.get<Movie>(
    MOVIE_DETAILS_PATH.replace("[id]", movieId) + `?api_key=${api_key}`
  );
