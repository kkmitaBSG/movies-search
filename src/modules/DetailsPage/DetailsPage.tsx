import React, { useEffect } from "react";
import { useDispatch, useSelector } from "react-redux";
import { useParams } from "react-router-dom";
import messages from "../../constants/messages";
import { MoviesActionType } from "../../store/types";
import { getMovieDetails } from "../../store/selectors";
import * as S from "./styles";

const DetailsPage = () => {
  const dispatch = useDispatch();
  const { id: movieId } = useParams();
  const details = useSelector(getMovieDetails(movieId!));

  useEffect(() => {
    dispatch({
      type: MoviesActionType.getMovieDetails,
      movieId,
    });
  }, [dispatch, movieId]);

  if (!details) {
    // TODO: skeleton loader
    return null;
  }

  const {
    title,
    backdrop,
    overview,
    production_countries,
    release_date,
    genres,
  } = details;

  return (
    <>
      <S.Poster src={backdrop}></S.Poster>
      <S.InfoWrapper>
        <S.LeftSide>
          <S.Title>{title}</S.Title>
          <S.Description>{overview}</S.Description>
        </S.LeftSide>

        <S.RightSide>
          <S.AdditionalInfo>
            <S.Name>{messages.AdditionalInfoProductionCountries}</S.Name>
            {production_countries?.[0]?.name}
          </S.AdditionalInfo>
          <S.AdditionalInfo>
            <S.Name>{messages.AdditionalInfoReleaseDate}</S.Name>
            {new Date(release_date!).getFullYear()}
          </S.AdditionalInfo>
          <S.AdditionalInfo>
            <S.Name>{messages.AdditionalInfoGenres}</S.Name>
            {genres?.map((genre) => genre.name).join(", ")}
          </S.AdditionalInfo>
        </S.RightSide>
      </S.InfoWrapper>
    </>
  );
};

export default DetailsPage;
