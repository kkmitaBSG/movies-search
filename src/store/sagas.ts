import { call, put, select, takeLatest } from "redux-saga/effects";
import { SagaIterator } from "redux-saga";
import { AnyAction } from "@reduxjs/toolkit";
import { MoviesActionType } from "./types";
import {
  getPopularMoviesNextPageSuccess,
  getPopularMoviesSuccess,
} from "./reducers/popularMoviesReducer";
import {
  getMovieDetails,
  getMoviesPage,
  getMoviesResults,
  getSearchFromLocationSearch,
  getSearchMoviesPage,
} from "./selectors";
import { getMovieDetailsSuccess } from "./reducers/movieDetailsReducer";
import {
  getSearchMoviesNextPageSuccess,
  getSearchMoviesSuccess,
} from "./reducers/searchMoviesReducer";
import * as api from "./requests";

export function* moviesWorker(): SagaIterator<void> {
  try {
    const moviesResults = yield select(getMoviesResults);

    if (moviesResults.length) {
      return;
    }

    const searchPhrase = yield select(getSearchFromLocationSearch);

    let response;
    if (searchPhrase) {
      response = yield call(api.getSearchMovies, searchPhrase, 1);
      yield put(getSearchMoviesSuccess(response.data));
    } else {
      response = yield call(api.getMovies, 1);
      yield put(getPopularMoviesSuccess(response.data));
    }
  } catch (err) {
    yield put({ type: MoviesActionType.getMoviesFail, message: err.message });
  }
}

export function* searchMoviesWorker(): SagaIterator<void> {
  try {
    const searchPhrase = yield select(getSearchFromLocationSearch);
    const response = yield call(api.getSearchMovies, searchPhrase, 1);
    yield put(getSearchMoviesSuccess(response.data));
  } catch (err) {
    yield put({ type: MoviesActionType.getMoviesFail, message: err.message });
  }
}

export function* moviesNextPageWorker(): SagaIterator<void> {
  try {
    const searchPhrase = yield select(getSearchFromLocationSearch);
    let response;
    if (searchPhrase) {
      const page: number = yield select(getSearchMoviesPage);
      response = yield call(api.getSearchMovies, searchPhrase, page + 1);
      yield put(getSearchMoviesNextPageSuccess(response.data));
    } else {
      const page: number = yield select(getMoviesPage);
      response = yield call(api.getMovies, page + 1);
      yield put(getPopularMoviesNextPageSuccess(response.data));
    }
  } catch (err) {
    yield put({ type: MoviesActionType.getMoviesFail, message: err.message });
  }
}

export function* movieDetailsWorker(action: AnyAction): SagaIterator<void> {
  try {
    const movieDetails = yield select(getMovieDetails(action.movieId));

    if (movieDetails) return;

    const response = yield call(api.getMovieDetails, action.movieId);

    yield put(getMovieDetailsSuccess(response.data));
  } catch (err) {
    yield put({ type: MoviesActionType.getMoviesFail, message: err.message });
  }
}

export default function* rootSaga() {
  yield takeLatest(MoviesActionType.getMovies, moviesWorker);
  yield takeLatest(MoviesActionType.getSearchMovies, searchMoviesWorker);
  yield takeLatest(MoviesActionType.getMoviesNextPage, moviesNextPageWorker);
  yield takeLatest(MoviesActionType.getMovieDetails, movieDetailsWorker);
}
