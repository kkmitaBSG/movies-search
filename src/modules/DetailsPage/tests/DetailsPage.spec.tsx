import * as reactRedux from "react-redux";
import renderer from "react-test-renderer";
import { MemoryRouter } from "react-router-dom";
import DetailsPage from "../DetailsPage";
import { MovieDetails } from "../../../store/types";

jest.mock("react-redux", () => ({
  ...jest.requireActual("react-redux"),
}));

const mockData: MovieDetails = {
  title: "title",
  backdrop: "backdrop",
  overview: "overview",
  production_countries: [{ name: "US" }],
  release_date: "2015",
  genres: [{ name: "commedy" }],
};

describe("DetailsPage", () => {
  const useSelectorMock = jest.spyOn(reactRedux, "useSelector");
  jest.spyOn(reactRedux, "useDispatch");

  test("render with movie details", () => {
    useSelectorMock.mockReturnValue(mockData);

    const component = renderer.create(
      <MemoryRouter>
        <DetailsPage />
      </MemoryRouter>
    );

    expect(component.toJSON()).toMatchSnapshot();
  });
});
