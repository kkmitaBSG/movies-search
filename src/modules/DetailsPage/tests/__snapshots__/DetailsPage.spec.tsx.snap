// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DetailsPage render with movie details 1`] = `
Array [
  <img
    className="sc-bcXHqe hfeTdG"
    src="backdrop"
  />,
  <div
    className="sc-gswNZR hAoysI"
  >
    <div
      className="sc-dkrFOg eRmXJd"
    >
      <div
        className="sc-eDvSVe fXFSTq"
      >
        title
      </div>
      <div
        className="sc-jSUZER coARuu"
      >
        overview
      </div>
    </div>
    <div
      className="sc-hLBbgP jBHWAK"
    >
      <div
        className="sc-gKPRtg dPDEXd"
      >
        <div
          className="sc-iBYQkv dwYivG"
        >
          Country:
        </div>
        US
      </div>
      <div
        className="sc-gKPRtg dPDEXd"
      >
        <div
          className="sc-iBYQkv dwYivG"
        >
          Release:
        </div>
        2015
      </div>
      <div
        className="sc-gKPRtg dPDEXd"
      >
        <div
          className="sc-iBYQkv dwYivG"
        >
          Genres:
        </div>
        commedy
      </div>
    </div>
  </div>,
]
`;
